aequilibrae/project/network/gmns_builder.py

```diff
@@ -0,0 +1,406 @@
+            letters = lt_name.lower() + lt_name.upper() + string.ascii_letters
+            letters = "".join([lt for lt in letters if lt not in all_types])
+
+            link_types = self.link_types
+            new_type = link_types.new(letters[0])
+            new_type.link_type = lt_name
+            new_type.description = "Link type from GMNS link table"
+            new_type.save()
+
+        # Creating modes list and saving modes to AequilibraE model
+
+        if gmns_modes_field in gmns_links_df.columns.to_list():
+            modes_list = gmns_links_df[gmns_modes_field].to_list()
+            if "" in modes_list:
+                modes_list = ["unspecified_mode" if x == "" else x for x in modes_list]
+
+        else:
+            modes_list = ["unspecified_mode" for _ in range(len(gmns_links_df))]
+
+        if use_group_path is not None:
+
+            use_group = pd.read_csv(use_group_path)
+            groups_dict = dict(zip(use_group.use_group, use_group.uses))
+            for k, use in groups_dict.items():
+                for group in list(groups_dict.keys()):
+                    if group in use:
+                        groups_dict[k] = use.replace(group, groups_dict[group])
+        else:
+            groups_dict = {}
+
+        char_replaces = {
+            "0": "zero",
+            "1": "one",
+            "2": "two",
+            "3": "three",
+            "4": "four",
+            "5": "five",
+            "6": "six",
+            "7": "seven",
+            "8": "eight",
+            "9": "nine",
+        }
+        pattern = re.compile("|".join(char_replaces.keys()))
+        modes_list = [
+            pattern.sub(lambda x: char_replaces[x.group()], s).replace("+", "").replace("-", "_") for s in modes_list
+        ]
+        mode_ids_list = [x for x in modes_list]
+
+        for mode in list(dict.fromkeys(modes_list)):
+
+            if mode in groups_dict.keys():
+                modes_gathered = [
+                    m.replace("+", "").replace("-", "_").replace("2", "two").replace("3", "three").replace(" ", "")
+                    for m in groups_dict[mode].split(sep=",")
+                ]
+                desc = use_group.loc[use_group.use_group == mode, "description"].item() + f" - GMNS use group: {mode}"
+
+            else:
+                modes_gathered = [m.replace("+", "").replace("-", "_").replace(" ", "") for m in mode.split(sep=",")]
+                if mode == "unspecified_mode":
+                    desc = "Mode not specified"
+                else:
+                    desc = "Mode from GMNS link table"
+
+            mode_to_add = ""
+            for m in modes_gathered:
+                letters = m.lower() + m.upper() + string.ascii_letters
+                all_modes = list(self.modes.all_modes())
+                letters = "".join([lt for lt in letters if lt not in all_modes and lt != "_"])
+
+                if m in [list(x.keys())[0] for x in p.parameters["network"]["modes"]]:
```

Comment:
m = m if m not  in [list(x.keys())[0] for x in p.parameters["network"]["modes"]] else m + "_gmns"